### PR TITLE
New Blog Post: The Human Bottleneck

### PR DESCRIPTION
**Pull request by oriondean**
Please add a direct link to your post here:

https://oriondean.github.io/blog/2026/05/12/the-human-bottleneck.html

Have you (please tick each box to show completion):

 - [x] Added your blog post to a single category?
 - [x] Added a brief summary for your post? Summaries should be roughly two sentences in length and give potential readers a good idea of the contents of your post.
 - [x] Checked that the build passes?
 - [x] Checked your spelling (you can use `npm install` followed by `npx mdspell "**/{FILE_NAME}.md" --en-gb -a -n -x -t` if that's your thing)
 - [x] Ensured that your author profile contains a profile image, and a brief description of yourself? (make it more interesting than just your job title!)
 - [x] Optimised any images in your post? They should be less than 100KBytes as a general guide.

Posts are reviewed / approved by your Regional Tech Lead.

### CODE REVIEW

File: _posts/2026-05-12-the-human-bottleneck.markdown [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+author: dkerr
+---
+
+The rapid acceleration of AI-augmented development has fundamentally shifted the software delivery bottleneck. As we write code exponentially faster, we are generating significantly more code that requires human review. As AI agents rapidly convert issues into potential solutions, the traditional pull-request queue swells, leaving the human reviewer as the primary constraint in the pipeline.
+
+There is more to gain than ever before from accelerating code delivery, tempting teams to reduce human oversight just to clear the pull-request queue. Top-down pressure often drives this push to eliminate the bottleneck, but doing so is fraught with risk from a multitude of standpoints: 
+
+### Security & Compliance
+  - **Security Vulnerabilities**: AI can confidently generate pristine-looking code that harbours subtle flaws (e.g., improper access controls or insecure dependencies).
```
Comment (amjones-scottlogic) [inline _posts/2026-05-12-the-human-bottleneck.markdown]: just an idea and im not sure how much we do it in our blogs but it could be good to link out to some of the big news cases of this and any of the ones below?

File: _posts/2026-05-12-the-human-bottleneck.markdown [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+There is more to gain than ever before from accelerating code delivery, tempting teams to reduce human oversight just to clear the pull-request queue. Top-down pressure often drives this push to eliminate the bottleneck, but doing so is fraught with risk from a multitude of standpoints: 
+
+### Security & Compliance
+  - **Security Vulnerabilities**: AI can confidently generate pristine-looking code that harbours subtle flaws (e.g., improper access controls or insecure dependencies).     
+  - **Compliance and Licensing**: Models may inadvertently introduce code that violates regulatory frameworks (e.g., GDPR) or breaches open-source licensing. 
+
+### System Integrity and Costs
+  - **Safety and Reliability**: AI lacks the contextual awareness to evaluate the blast radius of unhandled edge cases, risking real-world failures in critical systems. 
+  - **Operational Costs**: Unoptimised logic or inefficient database queries can slip through, leading to spikes in cloud infrastructure bills. Costs of using the AI models themselves are also notable.
```
Comment (amjones-scottlogic) [inline _posts/2026-05-12-the-human-bottleneck.markdown]: `nit-pick` The "Costs of using the AI models..." part just feels a little tagged on to the end. I think it might flow a bit more smoothly by saying something like "This is on top of the costs of using AI models int he first place" or ".. as well as the costs of using ai models..."

File: _posts/2026-05-12-the-human-bottleneck.markdown [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  - **Operational Costs**: Unoptimised logic or inefficient database queries can slip through, leading to spikes in cloud infrastructure bills. Costs of using the AI models themselves are also notable. 
+  - **Testing Blind Spots**: AI can write comprehensive-looking tests that perfectly validate its own flawed or incomplete logic, creating a false sense of security without verifying business requirements. 
+
+### Team Health and Architecture
+  - **Long-Term Maintainability**: Unreviewed code frequently introduces architectural drift and inconsistent patterns, compounding technical debt that slows future development. 
+  - **Stifling Developer Growth**: Undermining human review eliminates vital touchpoints where junior developers learn the codebase and senior developers provide mentorship. 
+  - **Erosion of Codebase Comprehension**: If AI writes and humans merely rubber-stamp, teams lose their mental model of the software architecture. This erosion of codebase knowledge limits the engineering team's ability to effectively prompt and guide AI agents on that same codebase in the longer term.
```
Comment (amjones-scottlogic) [inline _posts/2026-05-12-the-human-bottleneck.markdown]: is it also worth mentioning cognitive rot?

File: _posts/2026-05-12-the-human-bottleneck.markdown [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+There are, however, proven ways to relieve this pressure. Strategies like right-sizing risk, balancing short-term velocity with future friction, and automating parts of the review process can help manage the load. 
+
+## Right-Sizing Risk 
+
+Determining the appropriate level of human oversight requires an honest and accurate assessment of your project's risk appetite. While reducing oversight might be perfectly acceptable for a quick proof of concept, a throwaway internal script, or a low-stakes product lacking critical security concerns, applying that same leniency to a more critical system has dangerous consequences.
```
Comment (amjones-scottlogic) [inline _posts/2026-05-12-the-human-bottleneck.markdown]: `nit-pick` We dont directly state what right-sizing is here. The concept is explained but never fully confirmed. From reading this, I would presume its just catagorizing appetite for risks that leaves the questions of when and how does that fit into an SDLC for example.

I think something small like "Right-Sizing is the practice of restructuring a process to cover current needs" near the start would help the reader with a base for the rest of the section.

File: _posts/2026-05-12-the-human-bottleneck.markdown [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+### Shift-Left Testing and Pre-Commit Automation
+
+By pushing quality checks earlier in the development lifecycle, you can reduce the mental burden on human reviewers. Tools that automatically identify untested code and suggest edge-case tests within the PR can inherently boost your QA pipeline before the code is even merged. 
+
+You can shift the effort even further left by catching boilerplate errors and failing tests before the code is even committed. Leveraging git hooks to run automated linting, builds and tests on the developer's machine pre-commit ensures trivial mistakes are caught instantly, reducing broken code from reaching reviewer.
+
+### Stacked PRs 
+
+Because AI tools inherently drive up pull request sizes, breaking features into [smaller, dependent PRs](https://graphite.com/docs/best-practices-for-reviewing-stacks) prevents massive diffs from overwhelming reviewers. This enforces the author responsibility principle by requiring developers to curate changesets into more digestible chunks rather than offloading the cognitive burden onto the reviewer.
```
Comment (amjones-scottlogic) [inline _posts/2026-05-12-the-human-bottleneck.markdown:78]: Could be worth mentioning that AI tooling also makes it faster and easier to actually raise code for PR reducing some of the overheads of raising many smaller changes

File: _posts/2026-05-12-the-human-bottleneck.markdown [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+### Stacked PRs 
+
+Because AI tools inherently drive up pull request sizes, breaking features into [smaller, dependent PRs](https://graphite.com/docs/best-practices-for-reviewing-stacks) prevents massive diffs from overwhelming reviewers. This enforces the author responsibility principle by requiring developers to curate changesets into more digestible chunks rather than offloading the cognitive burden onto the reviewer. 
+
+## Conclusion 
+
+Ultimately, while AI widens the funnel at the top of the development lifecycle, the human bottleneck at the review stage remains a necessary, structural safeguard for long-term project health. 
+
+The goal shouldn't be to eliminate the human from the loop, but to elevate them. By right-sizing your risk, breaking down PRs, and enhancing review tooling we can effectively manage the human bottleneck without compromising code quality.
```
Comment (amjones-scottlogic) [inline _posts/2026-05-12-the-human-bottleneck.markdown]: Feels like this part just needs that little be extra to mention the gains. Something like:

"By right-sizing your risk, breaking down PRs, and enhancing review tooling we can effectively manage the human bottleneck, enabling the greater AI excellerated workflows, without compromising code quality.

File: _posts/2026-05-12-the-human-bottleneck.markdown [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+Next-generation tools address this fatigue by shifting from simple pattern matching to deep contextual understanding. Modern, AI-enhanced static analysis filters out noisy false positives using reachability analysis and contextual filtering, allowing reviewers to focus exclusively on actual, exploitable vulnerabilities.
+
+### Shift-Left Testing and Pre-Commit Automation
+
+By pushing quality checks earlier in the development lifecycle, you can reduce the mental burden on human reviewers. Tools that automatically identify untested code and suggest edge-case tests within the PR can inherently boost your QA pipeline before the code is even merged. 
+
+You can shift the effort even further left by catching boilerplate errors and failing tests before the code is even committed. Leveraging git hooks to run automated linting, builds and tests on the developer's machine pre-commit ensures trivial mistakes are caught instantly, reducing broken code from reaching reviewer.
```
Comment (jstrong-scottlogic) [inline _posts/2026-05-12-the-human-bottleneck.markdown]: Wonder if also worth mention post tool call hook checks for agents as well. e.g. it runs an edit file tool on a typescript file -> hook compiles, checks linting, and formatting before the AI does anything else.

**Review by jstrong-scottlogic — COMMENTED** [review]
Looks good. I would check about using pull-request or pull request throughout to make consistent and perhaps start using PR earlier.

File: _posts/2026-05-12-the-human-bottleneck.markdown (hunk 3 above)
Comment (oriondean) [reply _posts/2026-05-12-the-human-bottleneck.markdown]: Further down I linked to Addy Osmani's comprehension debt article - i could link up here too? Is there a similar article on cognitve rot you were referring to here?

File: _posts/2026-05-12-the-human-bottleneck.markdown (hunk 4 above)
Comment (oriondean) [reply _posts/2026-05-12-the-human-bottleneck.markdown]: Sounds good, I didn't want to get into the weeds of the practicalities of right-sizing as there's no one size fits all approach

File: _posts/2026-05-12-the-human-bottleneck.markdown (hunk 1 above)
Comment (oriondean) [reply _posts/2026-05-12-the-human-bottleneck.markdown]: I've added some additional literature to most of these points to support them

File: _posts/2026-05-12-the-human-bottleneck.markdown (hunk 3 above)
Comment (amjones-scottlogic) [reply _posts/2026-05-12-the-human-bottleneck.markdown]: Yeah that one is probably better, it was this one i was thinking of which i remembered being linked in the AI channel.

https://margaretstorey.com/blog/2026/02/09/cognitive-debt/

File: _posts/2026-05-12-the-human-bottleneck.markdown (hunk 2 above)
Comment (oriondean) [reply _posts/2026-05-12-the-human-bottleneck.markdown]: Updated

File: _posts/2026-05-12-the-human-bottleneck.markdown (hunk 5 above)
Comment (oriondean) [reply _posts/2026-05-12-the-human-bottleneck.markdown:78]: Added >The traditional overhead associated with managing numerous smaller changes is mitigated by AI tooling itself; by automatically generating PR descriptions and summarising changesets, AI makes it faster and easier to raise code for review, making stacked workflows much more viable for modern development teams.

File: _posts/2026-05-12-the-human-bottleneck.markdown (hunk 7 above)
Comment (oriondean) [reply _posts/2026-05-12-the-human-bottleneck.markdown]: Added 'or agent hooks' alongside git hooks with a link

File: _posts/2026-05-12-the-human-bottleneck.markdown (hunk 6 above)
Comment (oriondean) [reply _posts/2026-05-12-the-human-bottleneck.markdown]: Added something similar:  ...we can effectively manage the human bottleneck and reap the benefits of AI acceleration without compromising code quality.

**Review by jstrong-scottlogic — APPROVED** [review]

**Review by amjones-scottlogic — APPROVED** [review]
